`Pieces mismatches` when line includes multiple LoC joined by semicolon

Thanks for stack-data!

When multiple Python lines of code get joined together by semicolon assertion error `Pieces mismatches` is thrown.

Reproducer:

```python
import inspect
import stack_data

a = 1 ; b = ("""
   unhappy """
)
frame = inspect.currentframe()
frame_info = stack_data.FrameInfo(frame)
list(frame_info.lines)
```

Traceback:

```python
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    list(frame_info.lines)
  File "/site-packages/stack_data/utils.py", line 144, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/site-packages/stack_data/core.py", line 720, in lines
    pieces = self.included_pieces
  File "/site-packages/stack_data/utils.py", line 144, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/site-packages/stack_data/core.py", line 663, in included_pieces
    scope_pieces = self.scope_pieces
  File "/site-packages/stack_data/utils.py", line 144, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/site-packages/stack_data/core.py", line 603, in scope_pieces
    for piece in self.source.pieces
  File "/site-packages/stack_data/utils.py", line 144, in cached_property_wrapper
    value = obj.__dict__[self.func.__name__] = self.func(obj)
  File "/site-packages/stack_data/core.py", line 101, in pieces
    return list(self._clean_pieces())
  File "/site-packages/stack_data/core.py", line 125, in _clean_pieces
    raise AssertionError("Pieces mismatches: %s" % mismatches)
AssertionError: Pieces mismatches: [{4, 5}]
```

Originally reported in https://github.com/ipython/ipython/issues/13840.

Reproduced with:
- stack-data 0.6.1, 0.6.0, 0.5.1 (I did not test older)
- python 3.8.6 and 3.10.8

--- a/.github/workflows/pytest.yml
+++ b/.github/workflows/pytest.yml
@@ -2,7 +2,7 @@ name: Tests
 on: [push, pull_request]
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
     strategy:
       matrix:
         python-version: [3.6, 3.7, 3.8, 3.9, '3.10', 3.11-dev]

--- a/stack_data/core.py
+++ b/stack_data/core.py
@@ -124,6 +124,7 @@ class Source(executing.Source):
         for (start, end) in pieces[1:]:
             (last_start, last_end) = new_pieces[-1]
             if start < last_end:
+                assert start == last_end - 1
                 assert ';' in self.lines[start - 1]
                 new_pieces[-1] = (last_start, end)
             else:

--- a/stack_data/core.py
+++ b/stack_data/core.py
@@ -117,6 +117,19 @@ class Source(executing.Source):
             if end > start
         ]
 
+        # Combine overlapping pieces, i.e. consecutive pieces where the end of the first
+        # is greater than the start of the second.
+        # This can happen when two statements are on the same line separated by a semicolon.
+        new_pieces = pieces[:1]
+        for (start, end) in pieces[1:]:
+            (last_start, last_end) = new_pieces[-1]
+            if start < last_end:
+                assert ';' in self.lines[start - 1]
+                new_pieces[-1] = (last_start, end)
+            else:
+                new_pieces.append((start, end))
+        pieces = new_pieces
+
         starts = [start for start, end in pieces[1:]]
         ends = [end for start, end in pieces[:-1]]
         if starts != ends:

--- a/tests/samples/pieces.py
+++ b/tests/samples/pieces.py
@@ -94,3 +94,4 @@ def semicolons():
                        6)
             print(1, 2); print(3,
                   4); print(5, 6)
+            print(1, 2);print(3, 4);print(5, 6)

--- a/tests/samples/pieces.py
+++ b/tests/samples/pieces.py
@@ -79,3 +79,18 @@ class Foo(object):
     @property
     def foo(self):
         return 3
+
+
+# noinspection PyTrailingSemicolon
+def semicolons():
+    if 1:
+        print(1,
+         2); print(3,
+              4); print(5,
+                   6)
+        if 2:
+            print(1,
+             2); print(3, 4); print(5,
+                       6)
+            print(1, 2); print(3,
+                  4); print(5, 6)

--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -399,7 +399,20 @@ def test_pieces():
         ['class Foo(object):'],
         ['    @property',
          '    def foo(self):'],
-        ['        return 3']
+        ['        return 3'],
+        ['# noinspection PyTrailingSemicolon'],
+        ['def semicolons():'],
+        ['    if 1:'],
+        ['        print(1,',
+         '         2); print(3,',
+         '              4); print(5,',
+         '                   6)'],
+        ['        if 2:'],
+        ['            print(1,',
+         '             2); print(3, 4); print(5,',
+         '                       6)'],
+        ['            print(1, 2); print(3,',
+         '                  4); print(5, 6)'],
     ]
 
 

--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -413,6 +413,7 @@ def test_pieces():
          '                       6)'],
         ['            print(1, 2); print(3,',
          '                  4); print(5, 6)'],
+        ['            print(1, 2);print(3, 4);print(5, 6)']
     ]
 
 

